src/engines/kotor/userdata.cpp

```diff
@@ -0,0 +1,149 @@
+/* xoreos - A reimplementation of BioWare's Aurora engine
+ *
+ * xoreos is the legal property of its developers, whose names
+ * can be found in the AUTHORS file distributed with this source
+ * distribution.
+ *
+ * xoreos is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 3
+ * of the License, or (at your option) any later version.
+ *
+ * xoreos is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with xoreos. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file
+ *  Utility class for managing save data for kotor data.
+ */
+
+#include "src/common/readfile.h"
+#include "src/common/filepath.h"
+#include "src/common/filelist.h"
+#include "src/common/strutil.h"
+
+#include "src/aurora/talkman.h"
+
+#include "src/engines/kotor/userdata.h"
+#include "src/graphics/aurora/textureman.h"
+
+namespace Engines {
+
+namespace KotOR {
+
+UserData::UserData() : Engines::UserData("kotor"), _maxSaveId(1), _maxGameId(1) {
+	Common::FileList fileList;
+	fileList.addSubDirectories(_rootPath);
+	fileList.getSubListGlob("^[0-9]{9} - (Game[0-9]*|AUTOSAVE)$", false, fileList);
+
+	for (Common::FileList::const_iterator iter = fileList.begin(); iter != fileList.end(); iter++) {
+		Common::UString save = *iter;
+		if (
+			!Common::FilePath::isRegularFile(save + "/savenfo.res") ||
+			!Common::FilePath::isRegularFile(save + "/SAVEGAME.sav") ||
+			!Common::FilePath::isRegularFile(save + "/PARTYTABLE.res") ||
+			!Common::FilePath::isRegularFile(save + "/GLOBALVARS.res") ||
+			!Common::FilePath::isRegularFile(save + "/Screen.tga")
+		)
+			continue;
+
+		_saves.push_back(save);
+
+		save = Common::FilePath::getFile(save);
+		Common::UString id = save.substr(save.begin(), save.getPosition(6));
+		unsigned int value = 0;
+		Common::parseString(id, value);
+		_maxSaveId = MAX(_maxSaveId, value);
+
+		if (!save.contains("AUTOSAVE") && !save.contains("QUICKSAVE")) {
+			id = save.substr(save.getPosition(13), save.end());
+			Common::parseString(id, value);
```

Comment:
When the save name is malformed, `id` might be empty, in which case `Common::parseString()` throws. It might also also contain non-numerical data, in which case it also throws.

Reply:
The glob *might* already take care of it, but it's not obvious. Some comments might help.